nomad/variables_endpoint.go

```diff
@@ -113,56 +144,69 @@ func (sv *Variables) Apply(args *structs.VariablesApplyRequest, reply *structs.V
 	return nil
 }
 
-func svePreApply(sv *Variables, args *structs.VariablesApplyRequest, vd *structs.VariableDecrypted) (canRead bool, err error) {
+func hasReadPermission(aclObj *acl.ACL, namespace, path string) bool {
+	return aclObj.AllowVariableOperation(namespace,
+		path, acl.VariablesCapabilityRead, nil)
+}
 
-	canRead = false
-	var aclObj *acl.ACL
+func hasOperationPermissions(aclObj *acl.ACL, namespace, path string, op structs.VarOp) error {
 
-	// Perform the ACL resolution.
-	if aclObj, err = sv.srv.ResolveACL(args); err != nil {
-		return
-	} else if aclObj != nil {
-		hasPerm := func(perm string) bool {
-			return aclObj.AllowVariableOperation(args.Var.Namespace,
-				args.Var.Path, perm, nil)
-		}
-		canRead = hasPerm(acl.VariablesCapabilityRead)
+	hasPerm := func(perm string) bool {
+		return aclObj.AllowVariableOperation(namespace,
+			path, perm, nil)
+	}
 
-		switch args.Op {
-		case structs.VarOpSet, structs.VarOpCAS:
-			if !hasPerm(acl.VariablesCapabilityWrite) {
-				err = structs.ErrPermissionDenied
-				return
-			}
-		case structs.VarOpDelete, structs.VarOpDeleteCAS:
-			if !hasPerm(acl.VariablesCapabilityDestroy) {
-				err = structs.ErrPermissionDenied
-				return
-			}
-		default:
-			err = fmt.Errorf("svPreApply: unexpected VarOp received: %q", args.Op)
-			return
+	switch op {
+	case structs.VarOpSet, structs.VarOpCAS, structs.VarOpLockAcquire,
+		structs.VarOpLockRelease:
+		if !hasPerm(acl.VariablesCapabilityWrite) {
+			return structs.ErrPermissionDenied
 		}
-	} else {
-		// ACLs are not enabled.
-		canRead = true
+	case structs.VarOpDelete, structs.VarOpDeleteCAS:
+		if !hasPerm(acl.VariablesCapabilityDestroy) {
+			return structs.ErrPermissionDenied
+		}
+	default:
+		return fmt.Errorf("svPreApply: unexpected VarOp received: %q", op)
 	}
 
+	return nil
+}
+
+func canonicalizeAndValidate(args *structs.VariablesApplyRequest) error {
 	switch args.Op {
+	case structs.VarOpLockAcquire:
+		// If the variable doesn't have an ID, it means a lock is
+		// being created on a variable that doesn't exist, the variable needs to
+		// be created.
+		if args.Var.GetID() == "" {
+			args.Var.Canonicalize()
+		}
+
+		if err := args.Var.ValidateForLock(); err != nil {
+			return err
+		}
+
 	case structs.VarOpSet, structs.VarOpCAS:
 		args.Var.Canonicalize()
-		if err = args.Var.Validate(); err != nil {
-			return
-		}
+
+		return args.Var.Validate()
 
 	case structs.VarOpDelete, structs.VarOpDeleteCAS:
 		if args.Var == nil || args.Var.Path == "" {
-			err = fmt.Errorf("delete requires a Path")
-			return
+			return fmt.Errorf("delete requires a Path")
 		}
+
+	case structs.VarOpLockRelease:
+		if args.Var == nil || args.Var.Path == "" ||
+			args.Var.Namespace == "" || args.Var.Lock.ID == "" {
+			return errors.New("release requires all lock information")
+		}
+
+		return structs.ValidatePath(args.Var.Path)
```

Comment:
The variable already must exist in order for the operation to work. Do we need to validate it's path again here?

Reply:
If a user calls to acquire a lock on a variable that doesn't exit, it gets created, hence the path validation here.

Reply:
It's just a quick check to verify the provided path is not empty and is valid, to avoid making a call to the ss if we already know the path will throw an error.